lib/pbench/agent/tool_data_sink.py

```diff
@@ -381,216 +443,321 @@ def tm_log_capture(self):
         """tm_log_capture - capture all logs written by local and remote Tool
         Meisters through the Redis server into one file.
         """
+        self.logger.info("Running Tool Meister log capture ...")
+        # Create a separate logger so that the fetch_message() code only logs
+        # warnings and errors to stdout/stderr if problems when handling logs
+        # from remote Tool Meisters.
+        logger = logging.getLogger("tm_log_capture_thread")
+        logger.setLevel(logging.WARNING)
         tm_log_file = self.benchmark_run_dir / "tm" / "tm.logs"
         with tm_log_file.open("w") as fp:
             try:
-                for payload in self._tm_logging_chan:
-                    self.logger.debug("payload: %r", payload)
-                    assert (
-                        payload["channel"].decode("utf-8") == "tm-logging"
-                    ), f"{payload!r}"
-                    if payload["type"] == "unsubscribe":
-                        assert payload["pattern"] is None, f"{payload!r}"
-                        assert payload["data"] == 0, f"{payload!r}"
-                        break
-                    assert payload["type"] == "message", f"{payload!r}"
-                    try:
-                        log_str = payload["data"].decode("utf-8")
-                    except Exception:
-                        self.logger.warning(
-                            "log payload in message not UTF-8, %r", payload
-                        )
-                        continue
-                    else:
-                        fp.write(f"{log_str}\n")
-                        fp.flush()
+                for log_msg in self._to_logging_chan.fetch_message(logger):
+                    fp.write(f"{log_msg}\n")
+                    fp.flush()
             except redis.ConnectionError:
                 # We don't bother reporting any connection errors.
-                self._tm_logging_pubsub = None
+                pass
+            except ValueError as exc:
+                print(f"exc.args = {exc.args!r}", file=sys.stderr)
+                if exc.args[0] == "I/O operation on closed file.":
```

Comment:
Why do we need this again?

Reply:
When you ask yourself questions like this in public, it'd be nice if you answered them, too. ;-)

Also, ugh; reading exception arguments to make decisions isn't great. There's no more specific exception?

Reply:
Tracked in https://github.com/distributed-system-analysis/pbench/issues/2156.